### PR TITLE
fix: automate upgrade tests

### PR DESCRIPTION
**Pull request by mateusoliveira43**
## Why the changes were made

Automate upgrade tests process.

## How to test the changes made

Follow E2E testing documentation and run
```sh
make catalog-test-upgrade
TEST_UPGRADE=true make test-e2e
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -434,9 +435,9 @@ catalog-push: ## Push a catalog image.
 	$(MAKE) docker-push IMG=$(CATALOG_IMG)
 
 # A valid Git branch from https://github.com/openshift/oadp-operator
-PREVIOUS_CHANNEL ?= oadp-1.3
+PREVIOUS_CHANNEL ?= oadp-1.4
```
Comment (mateusoliveira43) [inline Makefile:438]: for each branch, we will have a different upgrade test (same foundation, but different DPA and logic, for example). Also, we will always test upgrades from next minor version

File: tests/e2e/upgrade_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+var _ = ginkgo.Describe("OADP upgrade scenarios", ginkgo.Ordered, func() {
+	// TODO do any clean up?
+	// var _ = ginkgo.AfterEach(func(ctx ginkgo.SpecContext) {
+	// 	tearDownCatalog()
+	// })
+
+	ginkgo.DescribeTable("Upgrade OADP channel tests",
+		func(scenario channelUpgradeCase) {
+			// Avoid DPA cleanup error on AfterSuite of upgrade only tests
```
Comment (mateusoliveira43) [inline tests/e2e/upgrade_suite_test.go:31]: TODO after adding logic, remove this

File: tests/e2e/sample-applications/upgrades/operator-install.yaml
Comment (mateusoliveira43) [inline tests/e2e/sample-applications/upgrades/operator-install.yaml:1]: Did not try to use this in code, but do not know if would make things much simpler

if an OperatorGroup already exists in namespace, things do not work...

File: tests/e2e/upgrade_suite_test.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+			dpaCR.CustomResource.Name = "dummyDPA"
+			dpaCR.CustomResource.Namespace = namespace
+
+			v1alpha1.AddToScheme(runTimeClientForSuiteRun.Scheme())
```
Comment (mateusoliveira43) [inline tests/e2e/upgrade_suite_test.go:35]: remove this after https://github.com/openshift/oadp-operator/pull/1447

File: tests/e2e/sample-applications/upgrades/operator-install.yaml
Comment (kaovilai) [reply tests/e2e/sample-applications/upgrades/operator-install.yaml:1]: > if an OperatorGroup already exists in namespace, things do not work...

correct. There can only be one operatorgroup per namespace.

[> If more than one OperatorGroup exists in a single namespace, any CSV created in that namespace will transition to a failure state with the reason TooManyOperatorGroups. CSVs in a failed state for this reason will transition to pending once the number of OperatorGroups in their namespaces reaches one.](https://docs.openshift.com/container-platform/4.2/operators/understanding_olm/olm-understanding-operatorgroups.html#:~:text=If%20more%20than,namespaces%20reaches%20one.)

Comment (mateusoliveira43) [reply tests/e2e/sample-applications/upgrades/operator-install.yaml:1]: deleted file for now and added shell commands in doc

File: tests/e2e/upgrade_suite_test.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+			}
+
+			log.Print("Creating Subscription oadp-operator")
+			subscription := v1alpha1.Subscription{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "oadp-operator",
+					Namespace: namespace,
+				},
+				Spec: &v1alpha1.SubscriptionSpec{
+					CatalogSource:          "oadp-operator-catalog-test-upgrade",
```
Comment (mateusoliveira43) [inline tests/e2e/upgrade_suite_test.go:63]: add ability to test prod operator

File: tests/e2e/upgrade_suite_test.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+						TargetNamespaces: []string{namespace},
+					},
+				}
+				err = runTimeClientForSuiteRun.Create(context.Background(), &operatorGroup)
+				gomega.Expect(err).To(gomega.BeNil())
+			}
+
+			subscriptionPackage := "oadp-operator"
+			subscriptionSource := "oadp-operator-catalog-test-upgrade"
+			if scenario.production {
```
Comment (weshayutin) [inline tests/e2e/upgrade_suite_test.go:53]: ah this is cool, that you have enabled this to test dev or production builds.  Nice

File: docs/developer/testing/test_oadp_version_upgrade.md [hunk 6]
```diff
@@ -26,6 +32,61 @@ Finally, after  creating a DPA, change channel of the installed operator to the
...
+  name: oadp-operator-group
+  namespace: <oadp-namespace>
+spec:
+  targetNamespaces:
+    - <oadp-namespace>
+```
+For example, run
+```sh
+export OADP_NAMESPACE=openshift-adp
+echo -e "kind: OperatorGroup\napiVersion: operators.coreos.com/v1\nmetadata:\n  name: oadp-operator-group\n  namespace: $OADP_NAMESPACE\nspec:\n  targetNamespaces:\n    - $OADP_NAMESPACE" | oc create -f -
```
Comment (weshayutin) [inline docs/developer/testing/test_oadp_version_upgrade.md:55]: there are many ways to create a yaml file with variables, this is fine. Not my preference but works :)

**Comment by weshayutin** [comment]
First pass at testing, looks GREAT!  Thank you @mateusoliveira43 

```
INFO[0002] Generating dockerfile                         bundles="[ttl.sh/oadp-operator-previous-bundle-2ccd054:1h ttl.sh/oadp-operator-bundle-2ccd054:1h]"
INFO[0002] writing dockerfile: ./index.Dockerfile1587693107  bundles="[ttl.sh/oadp-operator-previous-bundle-2ccd054:1h ttl.sh/oadp-operator-bundle-2ccd054:1h]"
INFO[0002] running docker build                          bundles="[ttl.sh/oadp-operator-previous-bundle-2ccd054:1h ttl.sh/oadp-operator-bundle-2ccd054:1h]"
INFO[0002] [docker build -f ./index.Dockerfile1587693107 -t ttl.sh/oadp-operator-catalog-2ccd054:1h .]  bundles="[ttl.sh/oadp-operator-previous-bundle-2ccd054:1h ttl.sh/oadp-operator-bundle-2ccd054:1h]"
docker push ttl.sh/oadp-operator-catalog-2ccd054:1h
The push refers to repository [ttl.sh/oadp-operator-catalog-2ccd054]
1fc0159cb68d: Pushed 
1a22300a0b9b: Pushed 
a33a02e436a5: Pushed 
eaa2ee7f1906: Pushed 
3f0dbb44ecc4: Pushed 
b336e209998f: Mounted from oadp-non-admin-0162550-4103 
f4aee9e53c42: Mounted from oadp-non-admin-0162550-4103 
1a73b54f556b: Mounted from oadp-non-admin-0162550-4103 
2a92d6ac9e4f: Mounted from oadp-non-admin-0162550-4103 
bbb6cacb8c82: Mounted from oadp-non-admin-0162550-4103 
ac805962e479: Mounted from oadp-non-admin-0162550-4103 
af5aa97ebe6c: Mounted from oadp-non-admin-0162550-4103 
4d049f83d9cf: Mounted from oadp-non-admin-0162550-4103 
945d17be9a3e: Mounted from oadp-non-admin-0162550-4103 
49626df344c9: Mounted from oadp-non-admin-0162550-4103 
f144bb4c7c7f: Pushed 
1h: digest: sha256:5dfc584784019a9954dd7ca6c625559525fd7e59b037d832d57a93cbbc94afd9 size: 3651
echo -e "apiVersion: operators.coreos.com/v1alpha1\nkind: CatalogSource\nmetadata:\n  name: oadp-operator-catalog-test-upgrade\n  namespace: openshift-marketplace\nspec:\n  sourceType: grpc\n  image: ttl.sh/oadp-operator-catalog-2ccd054:1h" | /home/whayutin/.local/bin/oc create -f -
catalogsource.operators.coreos.com/oadp-operator-catalog-test-upgrade created
chmod -R 777 test-upgrade && rm -rf test-upgrade && docker image rm catalog-test-upgrade
Untagged: catalog-test-upgrade:latest
Deleted: sha256:116919fbc2713f99616fe447e69a75bd9d8604cfbcc7ed42a2e114d1dd0465d4
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc whoami --show-server
https://c116-e.us-south.containers.cloud.ibm.com:30466
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ # after ui refresh the dev catalog showed up :)
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ # install from stable-1.4 oadp-1.4, success and I see oadp stable and v99.0 as well
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc create secret generic cloud-credentials --namespace openshift-adp --from-file cloud=/home/whayutin/OADP/SETUP/openshift_aws_credentials
error: failed to create secret secrets "cloud-credentials" already exists
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc create -f ~/OADP/SETUP/OADP_1.4/dpa.yaml 
dataprotectionapplication.oadp.openshift.io/dpa-sample created
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc get bsl
NAME           PHASE       LAST VALIDATED   AGE   DEFAULT
dpa-sample-1   Available   60s              78s   true
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ #now upgrade
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc get csv
NAME                   DISPLAY         VERSION   REPLACES   PHASE
oadp-operator.v1.4.0   OADP Operator   1.4.0                Succeeded
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ # sorry that ^ is previous :)
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ 
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ # upgrading 
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ 
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc get csv
NAME                    DISPLAY         VERSION   REPLACES               PHASE
oadp-operator.v99.0.0   OADP Operator   99.0.0    oadp-operator.v1.4.0   Succeeded
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ remote_velero version
Client:
	Version: 1.14.0
	Git commit: -
Server:
	Version: 1.14.0
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ # oadp-1.4 using the same velero image as master I think.. this is ok for this test
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc get bsl
NAME           PHASE       LAST VALIDATED   AGE    DEFAULT
dpa-sample-1   Available   9s               5m7s   true
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc get catalogsources.operators.coreos.com -n openshift-marketplace
NAME                                 DISPLAY               TYPE   PUBLISHER   AGE
certified-operators                  Certified Operators   grpc   Red Hat     6d23h
community-operators                  Community Operators   grpc   Red Hat     6d23h
oadp-operator-catalog-test-upgrade                         grpc               12m
redhat-marketplace                   Red Hat Marketplace   grpc   Red Hat     6d23h
redhat-operators                     Red Hat Operators     grpc   Red Hat     6d23h
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc get catalogsources.operators.coreos.com oadp-operator-catalog-test-upgrade -n openshift-marketplac
Error from server (NotFound): namespaces "openshift-marketplac" not found
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ oc get catalogsources.operators.coreos.com oadp-operator-catalog-test-upgrade -n openshift-marketplace -o yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  creationTimestamp: "2024-07-29T18:12:24Z"
  generation: 1
  name: oadp-operator-catalog-test-upgrade
  namespace: openshift-marketplace
  resourceVersion: "13904814"
  uid: 8ae4d8c8-76c0-4723-8064-7011c3db9479
spec:
  image: ttl.sh/oadp-operator-catalog-2ccd054:1h
  sourceType: grpc
status:
  connectionState:
    address: oadp-operator-catalog-test-upgrade.openshift-marketplace.svc:50051
    lastConnect: "2024-07-29T18:13:26Z"
    lastObservedState: READY
  registryService:
    createdAt: "2024-07-29T18:12:24Z"
    port: "50051"
    protocol: grpc
    serviceName: oadp-operator-catalog-test-upgrade
    serviceNamespace: openshift-marketplace
whayutin@thinkdoe:~/OPENSHIFT/git/OADP/oadp-operator$ 

```

**Review by weshayutin — APPROVED** [review]
/LGTM

**Review by shubham-pampattiwar — APPROVED** [review]

**Pull request merged**